### PR TITLE
Fix top banner ad label layout shift

### PR DESCRIPTION
**Pull request by arelra**
## What does this change?

Removes the layout shift caused by the advertisement label being dynamically inserted into the top banner space.

The change is to keep the label in the top banner template but not visible.

When the ad is rendered the label is made visible and its width set to centre it correctly.

DCR counterpart: https://github.com/guardian/dotcom-rendering/pull/2755

## Does this change need to be reproduced in dotcom-rendering ?

- [ ] No
- [X] Yes:  https://github.com/guardian/dotcom-rendering/pull/2755

## Screenshots

Before:

layout-shift-before-2

After:

layout-shift-after-01

## What is the value of this and can you measure success?

Improve core web vitals score

TODO: Lighthouse scores before and after

### Tested

- [X] Locally
- [X] On CODE (optional)

### CODE REVIEW

File: static/src/javascripts/projects/commercial/modules/dfp/render-advert-label.ts [hunk 1]
```diff
@@ -0,0 +1,109 @@
+/* eslint-disable @typescript-eslint/no-misused-promises
```
Comment (arelra) [inline static/src/javascripts/projects/commercial/modules/dfp/render-advert-label.ts:1]: This file has been converted to TypeScript.

Main for reference:

https://github.com/guardian/frontend/blob/main/static/src/javascripts/projects/commercial/modules/dfp/render-advert-label.js

File: static/src/javascripts/projects/commercial/modules/dfp/render-advert-label.ts [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+-- Fastdom measure and mutate signatures are Promise<void>
+-- Nested fastdom measure-mutate promises throw the error:
+-- "Promise returned in function argument where a void return was expected"
+*/
```
Comment (arelra) [inline static/src/javascripts/projects/commercial/modules/dfp/render-advert-label.ts:5]: This file has been converted to TypeScript.

Main for reference:

https://github.com/guardian/frontend/blob/main/static/src/javascripts/projects/commercial/modules/dfp/render-advert-label.js

File: static/src/javascripts/projects/commercial/modules/dfp/render-advert-label.ts [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+};
+
+const createAdLabel = () => {
+	const adLabel = document.createElement('div');
+	adLabel.className = 'ad-slot__label';
+	adLabel.innerHTML = 'Advertisement';
+	adLabel.appendChild(createAdCloseDiv());
+	return adLabel;
+};
+
```
Comment (arelra) [inline static/src/javascripts/projects/commercial/modules/dfp/render-advert-label.ts:38]: Main changes are to the function `renderAdvertLabel` to either render the top banner as a toggled element or as previously - inserting dynamically.

**Review by ioanna0 — APPROVED** [review]
great work 👍

File: common/app/views/fragments/commercial/topBanner.scala.html [hunk 4]
```diff
@@ -2,10 +2,13 @@
 
 @import views.support.Commercial.topAboveNavSlot
 
-<div class="@topAboveNavSlot.cssClasses(page.metadata)">
-  @fragments.commercial.standardAd(
-    "top-above-nav",
-    topAboveNavSlot.slotCssClasses,
-    topAboveNavSlot.adSizes
-  )
+<div>
```
Comment (ioanna0) [inline common/app/views/fragments/commercial/topBanner.scala.html:5]: I think we can remove the extra div wrapper

**Comment by mxdvl** [comment]
What does a toggleable advertisement label mean?

**Comment by arelra** [comment]
> What does a toggleable advertisement label mean?

It's an ad label whose visibility is toggled.
The previous method inserted the label dynamically which causes a layout shift.
Instead we keep the element in the DOM and hide/show when required.

**Comment by arelra** [comment]
Updated with:
- handling of ad slot label removal when ad slots are hidden i.e. when device < tablet https://github.com/guardian/frontend/pull/23613/commits/f04635f9c372093f5fa90a621a0ce00eabe2c621
- handling when the label should not be rendered e.g. when ad has an embedded label https://github.com/guardian/frontend/pull/23613/commits/14f7db9ab73b7a69fa86b284901d6a5331a775ae
- TS all the things https://github.com/guardian/frontend/pull/23613/commits/7f8bb1b49e28c3d8f4f6c8cd07a80a4af6eeb865 https://github.com/guardian/frontend/pull/23613/commits/17487fd289ea473bb52e38e78975537e20b39e6d

File: static/src/javascripts/projects/commercial/modules/remove-slots.spec.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+                    <div class="ad-slot__label ad-slot__label--toggle hidden">Advertisement</div>
+                    <div class="ad-slot__label ad-slot__label--toggle hidden">Advertisement</div>
+                    <div class="ad-slot__label">Advertisement</div>
+                    <div class="js-ad-slot"></div>
+                    <div class="js-ad-slot"></div>
+                </div>
+            `;
+        }
+        expect(Array.from(document.querySelectorAll(adSlotSelector)).length).toEqual(2);
+        expect(Array.from(document.querySelectorAll(toggledAdLabelSelector)).length).toEqual(2);
```
Comment (mxdvl) [inline static/src/javascripts/projects/commercial/modules/remove-slots.spec.js:25]: I think [`NodeList` has a `length` property](https://developer.mozilla.org/en-US/docs/Web/API/NodeList/length).

```suggestion
        expect(document.querySelectorAll(toggledAdLabelSelector).length).toEqual(2);
```

File: static/src/stylesheets/module/_adslot.scss [hunk 6]
```diff
@@ -60,6 +60,21 @@
     .ad-slot--sky & {
         width: $mpu-skyscraper-width;
     }
+
+    &--toggle {
+        margin: 0 auto;
+        @include mq($until: tablet) {
+            display: none;
+        }
```
Comment (mxdvl) [inline static/src/stylesheets/module/_adslot.scss:68]: Do they always show beyond tablet?

**Review by mxdvl — APPROVED** [review]
LGTM.

Some things I wonder about but are not blockers:
- When do we serve Ad slots that can be closed with the cross button?
- Why are ad labels never visible below tablet size?
- Why do we embed ad labels in fabric?

**Comment by arelra** [comment]
> LGTM.
> 
> Some things I wonder about but are not blockers:
> 
> * When do we serve Ad slots that can be closed with the cross button?
> * Why are ad labels never visible below tablet size?
> * Why do we embed ad labels in fabric?

1. Cross button

The close button is only for `.mobilesticky-container` ads.
However they are added to every ad label - but not shown which is wasteful.
I have a TODO to write a ticket to remove that button and decorate that specific ad in GAM, templates or the client side code.

2. Tablet size

Not sure of the historical reasons but the top banner is not shown for below tablet size:

https://github.com/guardian/frontend/blob/aff2baffd7fdd26a70fd711a52ddf891d0f36037/static/src/stylesheets/module/_adslot.scss#L140

Then the `commercial/dfp` at runtime actually removes all hidden ad slots at runtime.
Side note - I refactored the removal code to consolidate it, as it was duplicated in a couple of places.

3. Embedded labels

I don't know :)
I guess to give the commercial team some control over whether the label is shown.

I raised with @ashishpuliyel  that we should revisit the way labels are done - client side, templates and GAM.
If we can decorate all ads in GAM with the label and it would insert atomically with the ad, that would be ideal.
It would mean only one potential layout shift and we can delete all the client side code for labels.

File: static/src/stylesheets/module/_adslot.scss (hunk 6 above)
Comment (arelra) [reply static/src/stylesheets/module/_adslot.scss:68]: I believe so. I can't see any cases where we don't show beyond tablet in:

https://github.com/guardian/frontend/blob/aff2baffd7fdd26a70fd711a52ddf891d0f36037/static/src/stylesheets/module/_adslot.scss

File: static/src/javascripts/projects/commercial/modules/remove-slots.spec.js (hunk 5 above)
Comment (arelra) [reply static/src/javascripts/projects/commercial/modules/remove-slots.spec.js]: Thanks. Updated in a new commit

**Pull request merged**